DRMAAV2-14: Control and monitoring of Jobs

rampranesh opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[DRMAAV2-14](https://pbspro.atlassian.net/browse/DRMAAV2-14)**

#### Problem description
* Control and monitoring of jobs

#### Cause / Analysis
* NA

#### Solution description
* NA

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **unit test(s) to my commit**. *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why unit testing is not appropriate** for this case.
- [x] All new and existing automated tests have passed. 
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

vinodchitrali commented:
We need to design lazy fetch of data from DRMS(PBSPRo)

rampranesh commented:
Hi @vinodchitrali ,
      I have addressed your comments, could you please have a look at the changes.

Thanks,
Rampranesh M

arungrover commented on src/JobImpl.cpp:
```diff
@@ -0,0 +1,220 @@
+	if (_jobId.empty()) {
+		return _jobState;
+	}
+	PBSConnection pbsconn_(pbs_default(), 0, 0);
```
why do we store jobState separately? isn't jobState part of jobInfo?

rampranesh replied:
We have a method to query JobState alone, there is a possibility that application can call getState without calling getJobInfo. For this reason we have a copy of jobState separately.

arungrover commented on src/JobImpl.cpp:
```diff
@@ -0,0 +1,220 @@
+}
+
+const JobState& JobImpl::getState(string& subState) {
+	if (_jobId.empty()) {
```
I'd say you make default and copy constructors as private so that you don't have to deal with empty job ids

rampranesh replied:
Done added default and copy constructors private!

arungrover commented on src/JobImpl.cpp:
```diff
@@ -0,0 +1,220 @@
+}
+
+const JobTemplate& JobImpl::getJobTemplate(void) const {
+	return _jt;
```
is this a placeholder? will it be populated later on?

rampranesh replied:
JobTemplate is populated as part of the parameterized constructor, when a job is submitted the job id and the template used to submit the job are copied as part of the parameterised constructor. This method will return the populated JobTemplate object.

arungrover commented on src/JobImpl.cpp:
```diff
@@ -0,0 +1,220 @@
+	const Connection &pbsConnPoolObj_ = ConnectionPool::getInstance()->addConnection(pbsconn_);
+	DRMSystem *drms = Singleton<DRMSystem, PBSProSystem>::getInstance();
+	drms->suspend(pbsConnPoolObj_, *this);
+	ConnectionPool::getInstance()->returnConnection(pbsConnPoolObj_);
```
After looking at this code I looked at returnConnection and I think that returnConnection() needs to change too. It is doing a static cast to PBSConnection before switching the lists. Ideally it should not even be aware of PBSConnection.

rampranesh replied:
Done!

rampranesh commented on inc/JobImpl.h:
```diff
@@ -0,0 +1,218 @@
+	JobImpl(string& jobId_):_jobId(jobId_) {
+	}
+	/**
+         * Parameterised Constructor
```
Done!

rampranesh commented on inc/JobImpl.h:
```diff
@@ -0,0 +1,218 @@
+	/**
+         * Parameterised Constructor
+         */
+	JobImpl(string& jobId_, const JobTemplate& jt_):_jobId(jobId_), _jt(jt_) {
```
Done!

rampranesh commented on inc/JobImpl.h:
```diff
@@ -0,0 +1,218 @@
+	 */
+	const void populateJobInfo(void) const;
+
+	/**
```
Done!

rampranesh commented on inc/JobImpl.h:
```diff
@@ -0,0 +1,218 @@
+	 *
+	 * @return None
+	 */
+	virtual void suspend(void);
```
Done, added in all methods!

rampranesh commented on inc/JobTemplateAttrHelper.h:
```diff
@@ -65,6 +65,12 @@ class JobTemplateAttrHelper : public AttrHelper {
+	 * @brief parameterised constructor
+	 *
+	 */
+	JobTemplateAttrHelper(ATTRL* attrList_) : AttrHelper(attrList_) {
```
From our implementation point this can be const, but all our PBS ifl calls need struct attrl as input and not const struct attrl which is the reason we have left it this way.

rampranesh commented on src/JobImpl.cpp:
```diff
@@ -0,0 +1,220 @@
+		return;
+	}
+	_jobInfo.jobId = _jobId;
+	PBSConnection pbsconn_(pbs_default(), 0, 0);
```
First comment addressed, changed JobImpl to PBSJobImpl.
For the second comment, server name and port number are const data members of the class. Since they can only be initialized in constructors we have it this way.

rampranesh commented:
Hi @arungrover ,
    Thanks for your comments, I have addressed all your comments. Could you please have a look at these changes.

Thanks,
Rampranesh M

arungrover commented:
After another review please squash and rebase your branch

rampranesh commented:
Hi @arungrover,
  I have squashed and rebased my commit, could you please 
take this PR forward.
Thanks,
Rampranesh M

mike0042 commented on src/PBSJobImpl.cpp:
```diff
@@ -0,0 +1,214 @@
+					case 'F':
+						_jobInfo.jobState = DONE;
+						break;
+					default:
```
Shouldn't we be setting the job state to UNDETERMINED if we end up in the default block?
What about array jobs that are in state B? There are several other states not handled here.
Should you be handling cases for REQUEUED and REQUEUED_HELD here?
Same questions apply to PBSProSystem::state() as well.

rampranesh replied:
Hi @mike0042 ,
   Thanks for your comments, I have addressed all of them. DRMAA exposes getState method only for normal jobs, for array jobs there are no methods to query the state. So I have left the array job states in the implementation. Could you please have a look at these changes and let me know your comments.

Thanks,
Rampranesh M

rampranesh commented:
Hi @mike0042,
  I have rebased and squashed my commits could you please merge this PR.

Thanks,
Rampranesh M
